src/fmu/dataio/_metadata.py

```diff
@@ -80,9 +63,96 @@ def generate_meta_tracklog() -> list[meta.TracklogEvent]:
     ]
 
 
-@dataclass
-class MetaData:
-    """Class for sampling, process and holding all metadata in an ExportData instance.
+def _get_objectdata_provider(
+    object: types.Inferrable, dataio: ExportData, meta_existing: dict | None = None
+) -> ObjectDataProvider:
+    """Derive metadata for the object. Reuse metadata if existing"""
+    objdata = objectdata_provider_factory(object, dataio, meta_existing)
+    objdata.derive_metadata()
+    return objdata
+
+
+def _get_filedata_provider(
+    dataio: ExportData, objdata: ObjectDataProvider, fmudata: FmuProvider | None
+) -> FileDataProvider:
+    """Derive metadata for the file."""
+    filedata = FileDataProvider(
+        dataio=dataio,
+        objdata=objdata,
+        rootpath=dataio._rootpath,  # has been updated to case_path if fmurun
+        itername=fmudata.get_iter_name() if fmudata else "",
+        realname=fmudata.get_real_name() if fmudata else "",
+    )
+    filedata.derive_filedata()
+    return filedata
+
+
+def _compute_md5(
+    dataio: ExportData, objdata_extension: str, obj: types.Inferrable
+) -> str:
+    """Compute an MD5 sum using a temporary file."""
+    if objdata_extension.startswith("."):
+        with NamedTemporaryFile(buffering=0, suffix=objdata_extension) as tf:
+            logger.info("Compute MD5 sum for tmp file...: %s", tf.name)
+            return export_file_compute_checksum_md5(
+                obj=obj, filename=Path(tf.name), flag=dataio._usefmtflag
+            )
+    raise ValueError("An extension must start with '.'")
+
+
+def _get_meta_objectdata(
+    objdata: ObjectDataProvider,
+) -> meta.content.AnyContent | internal.UnsetAnyContent:
+    return (
+        internal.UnsetAnyContent.model_validate(objdata.metadata)
+        if objdata.metadata["content"] == "unset"
+        else meta.content.AnyContent.model_validate(objdata.metadata)
+    )
+
+
+def _get_meta_file(filedata: FileDataProvider, checksum_md5: str | None) -> meta.File:
+    return meta.File(
+        absolute_path=filedata.absolute_path,
+        relative_path=filedata.relative_path,
+        checksum_md5=checksum_md5,
+        relative_path_symlink=filedata.relative_path_symlink,
+        absolute_path_symlink=filedata.absolute_path_symlink,
+    )
+
+
+def _get_meta_access(access: dict | None) -> meta.SsdlAccess | None:
+    return meta.SsdlAccess.model_validate(access) if access else None
+
+
+def _get_meta_masterdata(masterdata: dict | None) -> meta.Masterdata | None:
+    return meta.Masterdata.model_validate(masterdata) if masterdata else None
```

Comment:
Maybe this `None` check should be done at the caller. Whenever possible we want to keep functions type stable, meaning they only and always return the same type. Python makes it very easy to return `None` to indicate something is not right but this ends up spreading complexity around because you have to add logic to handle this possibility to all callers

Reply:
Good point 🙂